ext/drbbdup/drbbdup.h

```diff
@@ -0,0 +1,334 @@
+
+/**
+ * \addtogroup drbbdup Basic Block Duplicator
+ */
+/*@{*/ /* begin doxygen group */
+
+/** Success code for each drbbdup operation */
+typedef enum {
+    DRBBDUP_SUCCESS,                       /**< Operation succeeded. */
+    DRBBDUP_ERROR_INVALID_PARAMETER,       /**< Operation failed: invalid parameter. */
+    DRBBDUP_ERROR_CASE_ALREADY_REGISTERED, /**< Operation failed: already registered. */
+    DRBBDUP_ERROR_CASE_LIMIT_REACHED,      /**< Operation failed: case limit reached. */
+    DRBBDUP_ERROR_ALREADY_INITIALISED,     /**< DRBBDUP can only be initialised once. */
+    DRBBDUP_ERROR,                         /**< Operation failed. */
+} drbbdup_status_t;
+
+/***************************************************************************
+ * User call-back functions
+ */
+
+/**
+ * Set up initial information related to a managing copies of a new basic block.
+ * A pointer-sized value indicating the default case encoding is returned.
+ * The boolean value written to \p enable_dups specifies whether code duplication should
+ * be done for this basic block. If false, the basic block is always executed under the
+ * default case and no duplications are made.
+ *
+ * Use drbbdup_register_case_value() to register other case encodings.
+ *
+ * @param[in] enable_dups indicates whether basic block code is duplicated to support
+ * multiple case instrumentation.
+ * @return the default case encoding.
+ */
+typedef uintptr_t (*drbbdup_set_up_bb_dups_t)(void *drbbdup_ctx, void *drcontext,
+                                              void *tag, instrlist_t *bb,
+                                              bool *enable_dups, void *user_data);
+
+/**
+ * Conducts a preliminary analysis of the basic block. The call-back is not called for
+ * each case, but once for the overall fragment. Therefore, computationally expensive
+ * analysis that only needs to be done once per fragment can be implemented by this
+ * call-back and made available to all cases of the basic block.
+ *
+ * The user can use thread allocation for storing the pre analysis result.
+ *
+ * @param[in] pre_analysis_data the pointer to store the result of the preliminary
+ * analysis.
+ * @return whether successful or an error code on failure.
+ */
+typedef void (*drbbdup_pre_analyse_bb_t)(void *drcontext, instrlist_t *bb,
+                                         void *user_data, void **pre_analysis_data);
+
+/**
+ * Destroys preliminary analysis data.
+ *
+ * The function is not invoked by drbbdup if pre analysis data was set to NULL by the
+ * drbbdup_pre_analyse_bb_t function.
+ *
+ * @param pre_analysis_data the preliminary analysis data to destroy.
+ */
+typedef void (*drbbdup_destroy_pre_analysis_t)(void *drcontext, void *user_data,
+                                               void *pre_analysis_data);
+
+/**
+ * Conducts an analysis on a basic block with respect to a case with encoding \p encoding.
+ * The result of the analysis needs to be stored to \p analysis_data.
```

Comment:
I did some rewording of this but essentially, we use void** because we are setting a void pointer that refers to the analysis data derived from this call-back.

Reply:
Done.